File: Week3/Day16/OddEvenLinkedList/cpp/OddEvenLinkedList.cpp
// Source : https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/536/week-3-may-15th-may-21st/3331/
// Author : Shubhanshu Singh
// Date   : 17-05-2020


// Time Complexity: O(n)
// Space Complexity: O(1)


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        
        if(!head)
            return head;
        
        ListNode *odd = head;
        ListNode *evenHead = head->next;
        ListNode *even = head->next;
        
        while(even && even->next)
        {
            odd->next = odd->next->next;
            even->next = even->next->next;
            
            even = even->next;
            odd = odd->next;
        }
        
        odd->next = evenHead;
        
        return head;
    }
};

File: Week3/Day16/OddEvenLinkedList/python/OddEvenLinkedList.py
## Source : https://leetcode.com/explore/challenge/card/may-leetcoding-challenge/536/week-3-may-15th-may-21st/3331/
## Author : Shubhanshu Singh
## Date   : 17-05-2020


## Time Complexity: O(n)
## Space Complexity: O(1)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        
        if(not head):
            return head
        
        odd = head
        evenHead = head.next
        even = head.next
        
        while(even and even.next):
            odd.next = odd.next.next
            even.next = even.next.next
            even = even.next
            odd = odd.next
        
        odd.next = evenHead
        
        return head
        